Patrol CLI sometimes throws "Connection closed before full header was received" error

### Is there an existing issue for this?

- [X] I have searched the [existing issues](https://github.com/leancodepl/patrol/issues)

### Steps to reproduce

1. You can run this workflow: https://github.com/leancodepl/patrol/actions/workflows/test-ios-simulator.yaml

### Actual results

I have noticed it only while testing on iOS simulator but I'm not sure it can't happen while testing on Android.
Two out of about thirty times `patrol build` or `patrol test` is throwing an error: `Connection closed before full header was received` and command fails to run or build tests.
Check run here: https://github.com/leancodepl/patrol/actions/runs/4990196745/jobs/8935097806#step:8:20

### Logs

<details>
<summary>Logs</summary>

```
Connection closed before full header was received
#0      IOClient.send (package:http/src/io_client.dart:90:7)
<asynchronous suspension>
#1      BaseClient._sendUnstreamed (package:http/src/base_client.dart:93:32)
<asynchronous suspension>
#2      _withClient (package:http/http.dart:164:12)
<asynchronous suspension>
#3      PubUpdater._getPackageInfo (package:pub_updater/src/pub_updater.dart:68:22)
<asynchronous suspension>
#4      PubUpdater.getLatestVersion (package:pub_updater/src/pub_updater.dart:44:25)
<asynchronous suspension>
#5      PatrolCommandRunner._checkForUpdate (package:patrol_cli/src/runner/patrol_command_runner.dart:365:27)
<asynchronous suspension>
#6      PatrolCommandRunner.runCommand (package:patrol_cli/src/runner/patrol_command_runner.dart:306:7)
<asynchronous suspension>
#7      PatrolCommandRunner.run (package:patrol_cli/src/runner/patrol_command_runner.dart:258:18)
<asynchronous suspension>
#8      patrolCommandRunner (package:patrol_cli/src/runner/patrol_command_runner.dart:66:[20](https://github.com/leancodepl/patrol/actions/runs/4990196745/jobs/8935097806#step:8:21))
<asynchronous suspension>
#9      main (file:///Users/runner/work/patrol/patrol/packages/patrol_cli/bin/main.dart:6:20)
<asynchronous suspension>

Error: Process completed with exit code 1.
```

</details>


### Patrol version

patrol: ^1.2.0-dev.3
patrol_cli: ^1.1.9

### Patrol Doctor output

<details>
<summary>Patrol Doctor output</summary>

```
I don't need a doctor
```

</details>


### Flutter Doctor output

<details>
<summary>Flutter Doctor output</summary>

```
I don't have
```

</details>

--- a/packages/patrol_cli/lib/src/analytics/analytics.dart
+++ b/packages/patrol_cli/lib/src/analytics/analytics.dart
@@ -2,7 +2,6 @@
 import 'dart:convert';
 import 'dart:io' as io;
 
-import 'package:ci/ci.dart' as ci;
 import 'package:file/file.dart';
 import 'package:http/http.dart' as http;
 import 'package:patrol_cli/src/base/constants.dart';
@@ -43,11 +42,13 @@ class Analytics {
     required Platform platform,
     http.Client? httpClient,
     FlutterVersionGetter getFlutterVersion = _getFlutterVersion,
+    required bool isCI,
   })  : _fs = fs,
         _platform = platform,
         _httpClient = httpClient ?? http.Client(),
         _postUrl = _getAnalyticsUrl(measurementId, apiSecret),
-        _flutterVersion = getFlutterVersion();
+        _flutterVersion = getFlutterVersion(),
+        _isCI = isCI;
 
   final FileSystem _fs;
   final Platform _platform;
@@ -56,6 +57,7 @@ class Analytics {
   final String _postUrl;
 
   final FlutterVersion _flutterVersion;
+  final bool _isCI;
 
   /// Sends an event to Google Analytics that command [name] run.
   ///
@@ -97,10 +99,8 @@ class Analytics {
     );
   }
 
-  bool get isCI => ci.isCI;
-
   bool get enabled {
-    if (isCI) {
+    if (_isCI) {
       return false;
     }
 

--- a/packages/patrol_cli/lib/src/runner/patrol_command_runner.dart
+++ b/packages/patrol_cli/lib/src/runner/patrol_command_runner.dart
@@ -3,6 +3,7 @@ import 'dart:io' show ProcessSignal, stdin;
 import 'package:adb/adb.dart';
 import 'package:args/args.dart';
 import 'package:args/command_runner.dart';
+import 'package:ci/ci.dart' as ci;
 import 'package:cli_completion/cli_completion.dart';
 import 'package:dispose_scope/dispose_scope.dart';
 import 'package:file/file.dart';
@@ -344,6 +345,11 @@ Ask questions, get support at https://github.com/leancodepl/patrol/discussions''
   }
 
   bool _wantsUpdateCheck(String? commandName) {
+    if (ci.isCI) {
+      // We don't want to check for updates on CI because of #1282
+      return false;
+    }
+
     if (commandName == 'update' || commandName == 'doctor') {
       return false;
     }

--- a/packages/patrol_cli/lib/src/runner/patrol_command_runner.dart
+++ b/packages/patrol_cli/lib/src/runner/patrol_command_runner.dart
@@ -37,6 +37,7 @@ Future<int> patrolCommandRunner(List<String> args) async {
   const fs = LocalFileSystem();
   const platform = LocalPlatform();
   final processManager = LoggingLocalProcessManager(logger: logger);
+  final isCI = ci.isCI;
 
   final runner = PatrolCommandRunner(
     pubUpdater: pubUpdater,
@@ -48,8 +49,10 @@ Future<int> patrolCommandRunner(List<String> args) async {
       apiSecret: _gaApiSecret,
       fs: fs,
       platform: platform,
+      isCI: isCI,
     ),
     processManager: processManager,
+    isCI: isCI,
   );
 
   if (!platform.environment.containsKey('PATROL_MIGRATED')) {
@@ -90,6 +93,7 @@ class PatrolCommandRunner extends CompletionCommandRunner<int> {
     required ProcessManager processManager,
     required Analytics analytics,
     required Logger logger,
+    required bool isCI,
   })  : _platform = platform,
         _pubUpdater = pubUpdater,
         _fs = fs,
@@ -97,6 +101,7 @@ class PatrolCommandRunner extends CompletionCommandRunner<int> {
         _processManager = processManager,
         _disposeScope = DisposeScope(),
         _logger = logger,
+        _isCI = isCI,
         super(
           'patrol',
           'Tool for running Flutter-native UI tests with superpowers',
@@ -218,6 +223,7 @@ class PatrolCommandRunner extends CompletionCommandRunner<int> {
 
   final DisposeScope _disposeScope;
   final Logger _logger;
+  final bool _isCI;
 
   Future<void> dispose() async {
     try {
@@ -318,7 +324,7 @@ Ask questions, get support at https://github.com/leancodepl/patrol/discussions''
   void printUsage() => _logger.info(usage);
 
   void _handleAnalytics() {
-    if (_analytics.firstRun && !_analytics.isCI) {
+    if (_analytics.firstRun && !_isCI) {
       _logger.info(
         '''
 \n
@@ -345,7 +351,7 @@ Ask questions, get support at https://github.com/leancodepl/patrol/discussions''
   }
 
   bool _wantsUpdateCheck(String? commandName) {
-    if (ci.isCI) {
+    if (_isCI) {
       // We don't want to check for updates on CI because of #1282
       return false;
     }

--- a/packages/patrol_cli/test/analytics/analytics_test.dart
+++ b/packages/patrol_cli/test/analytics/analytics_test.dart
@@ -36,6 +36,7 @@ void main() {
         platform: fakePlatform('/Users/john'),
         httpClient: httpClient,
         getFlutterVersion: FlutterVersion.test,
+        isCI: false,
       );
     });
 

--- a/packages/patrol_cli/test/commands/patrol_command_runner_test.dart
+++ b/packages/patrol_cli/test/commands/patrol_command_runner_test.dart
@@ -40,6 +40,7 @@ void main() {
         fs: MemoryFileSystem.test(),
         analytics: MockAnalytics(),
         logger: logger,
+        isCI: false,
       );
     });
 

